python/tests/test_dict_encoding.py

```diff
@@ -404,3 +487,27 @@ def test_provenances(self):
             "provenances",
             ["record", "record_offset", "timestamp", "timestamp_offset"],
         )
+
+    def test_top_level_metadata(self):
+        tables = get_example_tables()
+        d = tables.asdict()
+        d["metadata"] = None
```

Comment:
We also want to test what happens when we ``del d["metadata"]``, right?

Reply:
https://github.com/tskit-dev/tskit/pull/687/commits/e59142657d1e9a27d15f39e46ed18d52bef98d9d

Reply:
I don't think this is what we want. The dict encoding is the de-facto interchange format, and so we must be robust to reading in data from older versions of msprime and tsinfer (etc). Erroring out if there's an optional key missing is something we just can't do. The semantics are pretty straightforward (and already handled in the maps-to-None case), so I think it's an easy fix. All we need to do is update ``get_table_dict_value`` to return None when the key is missing and required is false.

I just tried it out, and it seems to work fine. There's a little bit of breakage, but should only be a few minutes to clear up.

This will also simplify some of the functions I pointed out below, where we're adding in keys mapping to None when the parameters are missing.

Reply:
I've has a go in https://github.com/tskit-dev/tskit/pull/687/commits/682b97da754cb07b00969d8d75c6b06d0314976f Missing used to be a `ValueError` and `None` a `TypeError` I've had to harmonise this to `TypeError` we hit other exceptions that have `TypeError` such as those for a missing pair in a ragged set when letting `None` fall through.